Data Classifications for resources and attributes (#187)
open-telemetry/oteps: text/0187-data-classification.md

# Introduce Data Classification for Telemetry

Adding optional classification to the attributes and resources to support simplified processing of data.

## Motivation

As the scope of Observability changes to include user monitoring and analytics (Real User Monitoring for now), ensuring that data is handled correctly is becoming more problematic as the problem space changes to include more. The need for pre processing data in order that high cardinality, sensitive user data, and handling sparse data impacts the reliability and usability of the data. Furthermore, some vendors are not able to selectively remove sensitive user data or some agreements with customers mandate that sensitive data not to leave the internal edge. Moreover, attribute values that are known high cardinality (for example a container id within a cloud based environment) that are expensive to store within observability systems, making it easier to omit attributes as part of exporter or collector configuration can greatly reduce cost of the observability suite.; not all attributes and resource data hold the same value or have the same processing requirements.

> Updating the SDK to include the convention of data classification would mean that users can extend their current telemetry via an opt in process to enable enhanced controls over the telemetry data without impacting the developer experience. Instrumentation authors are not limited by what attributes they can use when developing their telemetry and not a hindrance for existing standards within Open Telemetry. Organisations can define what resource classifications are allowed to be processed by an observability suite, what is subject to additional processing, and understand the types of data being sent. For example, all high cardinality data can be sent to on prem short term storage and then forwarded to a vendor with high cardinality attributes removed so that it can be viewed over a longer time period to show trends.

Reviewer: ```suggestion
Updating the SDK to include the convention of data classification would mean that users can extend their current telemetry via an opt in process to enable enhanced controls over the telemetry data without impacting the developer experience. Instrumentation authors are not limited by what attributes they can use when developing their telemetry and not a hindrance for existing standards within OpenTelemetry. Organisations can define what resource classifications are allowed to be processed by an observability suite, what is subject to additional processing, and understand the types of data being sent. For example, all high cardinality data can be sent to on prem short term storage and then forwarded to a vendor with high cardinality attributes removed so that it can be viewed over a longer time period to show trends.
```

Having the ability to set data classification on resources means that:

Known high cardinality attributes can be filtered out to reduce the cost of Observability suites.

> Attributes containing User Generated Content, Personal Data or sensitive data can be rerouted into a slower processing queue to redact, anonymise, or drop resource values.

Reviewer: ```suggestion
Attributes containing User Generated Content (UGC), Personal Data (PD) or sensitive data can be rerouted into a slower processing queue to redact, anonymise, or drop resource values.
```

Associate resource telemetry to Service Level Objective / Service Level Indicator (SLO/SLI) so that observability systems can automatically identify data that should be alerted on.  

Ensure that sparse data is not sampled out by mistake

This approach can also be extended by vendors to enable more advanced features in their tool such as high resolution, short retention, data residency, or an endless list of features that can be adopted.

## Explanation

> A service that adopts using resource classifications allows the open telemetry exporters to route data, remove problematic attributes, or filter values being sent to downstream systems. Using classifications simplify the existing processes since it does not use a lookup table that needs to be maintained, regular expressions needing to be maintained, required user intervention of what is acceptable or not, or needing to validate the entire resource object thus meaning no performance impact.  

Reviewer: ```suggestion
A service that adopts using resource classifications allows the OpenTelemetry exporters to route data, remove problematic attributes, or filter values being sent to downstream systems. Using classifications simplify the existing processes since it does not use a lookup table that needs to be maintained, regular expressions needing to be maintained, required user intervention of what is acceptable or not, or needing to validate the entire resource object thus meaning no performance impact.  
```

    ctx, span := otel.GetTraceProvider().Tracer("otel-go-middleware").Start(ctx, privacy.Sanitize(r.URL.Path()),
        trace.WithAttibutes(
          attribute.String("net.peer.ip", r.SourceIP).Classify(attribute.PII, attribute.HighCardinality),
          attribute.String("http.url", r.URL).Classify(attribute.Sensitive)
          attribute.String("http.user_agent", r.UserAgent).Classify(attributes.Sensitive, attributes.HighCardinality)
          attribute.Int("retry.amount", attempts), // Classifications are considered optional by the SDK 
          // ... Other default attributes as defined by the semantic convention
        ), 
      )
    defer span.Stop()
  
    interceptor := otel.NewResponseInteceptor(rw)
    
    defer func() {
      if interceptor.IsError() {
        span.SetAttributes(
          attributes.Error(interceptor.Error()).Classify(attribute.PreserveValue)
        )
      }
    }()
    
    return next.Serve(interceptor, r)
  }))
}
```

Once classification have been added to the attribute keys, a tracing sampler can

```go
func (ts TailBasedSampler) Sample(spans trace.Spans) (export trace.Spans) {
  for _, s := range spans {

>     if classification.Match(s.Classification(), classification.Ephermal) {

Reviewer: ```suggestion
    if classification.Match(s.Classification(), classification.Ephemeral) {
```

      export.Append(s)

>       // Always retain spans that contain ephermal attributes

Reviewer: ```suggestion
      // Always retain spans that contain ephemeral attributes
```

A staring list can look like the following :

| Classification                     | Purpose                                                                                               | BitShift  | Hint Value (base 16)  | bit mask             |
|----------------------------------- |------------------------------------------------------------------------------------------------------ |---------- |---------------------- |--------------------- |
| No Value                           | Is the default value when un set                                                                      | 0 << 0    | 0x0000                | 0000 0000 0000 0000  |
| Ephemeral                          | The attributes that are short lived and have high potential to change over extended periods of time.  | 1 << 0    | 0x0001                | 0000 0000 0000 0001  |

> | High Cardinality                   | The value is an unbounded set                                                                         | 1 << 1    | 0x0002                | 0000 0000 0000 0001  |

Reviewer 2: ```suggestion
| High Cardinality                   | The value is an unbounded set                                                                         | 1 << 1    | 0x0002                | 0000 0000 0000 0010  |
```

| Sensitive Value                    | The value MAY contain information that requires sanitisation                                          | 1 << 2    | 0x0004                | 0000 0000 0000 0100  |
| Personal Identifiable Information  | The value DOES contain Personal Identifiable Information                                              | 1 << 3    | 0x0008                | 0000 0000 0000 1000  |
| User Generated Content             | The value DOES contain User Generated Content                                                         | 1 << 4    | 0x0010                | 0000 0000 0001 0000  |

> | Service Level Objective            | The value is used to track a service level objective                                                  | 1 << 5    | 0x0020                | 0000 0000 0010 0000  |

Reviewer: I'm not sure I would include this one, at least not for the first iteration. My impression is that this wouldn't have the same audience as the other values.

Author: For Service Level Object? Or classifications themselves?

Reviewer: I mean only the SLO classification.

  - This only works with USERs implementing their code using OTEL instrumentation.
- Converting from non OTLP formats to OTLP would need depend on processors / converters to support this, however doing so will likely impact performance.
- Managing classifications within the semantic convention has not be explored
- The tolerance of what is considered a given classification could change between organisations

## Prior art and alternatives

There was a conversation that was related here on attributes that should be subject to data regulations Proposal: Add Sensitive Data LabelsOPEN , however, this approach would superseded what is mentioned there and be applied more broadly to the project.

The alternatives that were considered are:

- The Semantic Convention defines classification for known attributes
  - this was no considered heavily since it misses USERS (developers) whom are using the SDK natively and then would need to manage their own Semantic Convention that would also need to be understood by the down stream processors and vendors
    - Does not account for version drift with a breaking change involved
    - Does a processor need to know all versions?

>     - Does that mean values can not be changed in the semantic convention?

Reviewer 3: Another downside of doing it in semantic conventions goes to your point above that different organizations will have potentially different classifications. Reaching a common agreement on classification for the semantic conventions would require lengthy discussion, with a good chance of many attributes not reaching a consensus decision on a classification

    - Unneeded work that should be avoided
    - Limits what USERs can define their attributes keys
  - The ownership is then moved to processors and vendors to set the definition
    - Goes against the projects goal of being vendor neutral
    - Each vendor processor could define different values (same as above)

## Open questions

Does the semantic convention define what hints are automatically set for exported attributes?

Can a Service Owner change the definitions of classifications without needed to modify instrumentation code?

How would this work with the semantic convention?

What definitions should be used for the classifications so that the values are set subjectively?

What extra work is missing to make classifications easy to manage as part of the semantic convention?

## Future possibilities

Some ideas of future possibilities have been eluded to throughout the proposal, the potential changes that would be of great impact are:

- Vendor based classification controls
- Vendor neutral data protection mechanisms
- Enforcing Data Regulations on telemetry
- Classification based routing
  - An exporter can only with certain set of classifications

> Thinking more broadly, it would mean that Open Telemetry could easily extend to support analytical based telemetry on user interactions while adhering to data regulation or organisational policies.

Reviewer: ```suggestion
Thinking more broadly, it would mean that OpenTelemetry could easily extend to support analytical based telemetry on user interactions while adhering to data regulation or organisational policies.
```
